test/e2e/instrumentation/monitoring/stackdriver_metadata_agent.go

```diff
@@ -0,0 +1,138 @@
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce", "gke")
+	})
+
+	f := framework.NewDefaultFramework("stackdriver-monitoring")
+	var kubeClient clientset.Interface
+
+	It("should run Stackdriver Metadata Agent [Feature:StackdriverMetadataAgent]", func() {
+		kubeClient = f.ClientSet
+		testAgent(f, kubeClient)
+	})
+})
+
+func testAgent(f *framework.Framework, kubeClient clientset.Interface) {
+	projectId := framework.TestContext.CloudConfig.ProjectID
+	resourceType := "k8s_container"
+	uniqueContainerName := fmt.Sprintf("test-container-%v", time.Now().Unix())
+	endpoint := fmt.Sprintf(
+		"https://stackdriver.googleapis.com/v1beta2/projects/%v/resourceMetadata?filter=resource.type%%3D%v+AND+resource.label.container_name%%3D%v",
+		projectId,
+		resourceType,
+		uniqueContainerName)
+
+	oauthClient, err := google.DefaultClient(context.Background(), MonitoringScope)
+	if err != nil {
+		framework.Failf("Failed to create oauth client: %s", err)
+	}
+
+	// Create test pod with unique name.
+	framework.CreateExecPodOrFail(kubeClient, f.Namespace.Name, uniqueContainerName, func(pod *v1.Pod) {
+		pod.Spec.Containers[0].Name = uniqueContainerName
+	})
+	defer kubeClient.CoreV1().Pods(f.Namespace.Name).Delete(uniqueContainerName, &metav1.DeleteOptions{})
+
+	// Wait a short amount of time for Metadata Agent to be created and metadata to be exported
+	time.Sleep(120 * time.Second)
+
+	resp, err := oauthClient.Get(endpoint)
+	if err != nil {
+		framework.Failf("Failed to call Stackdriver Metadata API %s", err)
+	}
+	if resp.StatusCode != 200 {
+		framework.Failf("Stackdriver Metadata API returned error status: %s", resp.Status)
+	}
+	metadataAPIResponse := new(bytes.Buffer)
+	_, err = metadataAPIResponse.ReadFrom(resp.Body)
+	if err != nil && err != io.EOF {
+		framework.Failf("Failed to read response from Stackdriver Metadata API: %s", err)
+	}
+
+	exists, err := verifyPodExists(metadataAPIResponse.Bytes(), uniqueContainerName)
+	if err != nil {
+		framework.Failf("Failed to process response from Stackdriver Metadata API: %s", err)
+	}
+	if !exists {
+		framework.Failf("Missing Metadata for container %q", uniqueContainerName)
+	}
+}
+
+type Metadata struct {
+	Results []map[string]interface{}
+}
+
+type Resource struct {
+	resourceType   string
+	resourceLabels map[string]string
+}
+
+func verifyPodExists(response []byte, containerName string) (bool, error) {
+	var metadata Metadata
+	err := json.Unmarshal(response, &metadata)
+	if err != nil {
+		return false, fmt.Errorf("Failed to unmarshall: %s", err)
+	}
+
+	for _, result := range metadata.Results {
+		resource, err := parseResource(result["resource"])
+		if err != nil {
+			return false, fmt.Errorf("No 'resource' label: %s", err)
+		}
+		if resource.resourceType == "k8s_container" &&
+			resource.resourceLabels["container_name"] == containerName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func parseResource(result interface{}) (*Resource, error) {
```

Comment:
Two things:
- You're not handling errors in this function. What if some type casts fail or labels/type field is missing?
- "result" is hardly an appropriate name for the function argument here.

Reply:
Done